### PR TITLE
Update to use NullableArrays and CategoricalArrays

### PR DESCRIPTION
**Pull request by randy3k**
TODO

- [x] Handle ordered factor
- [x] Optionally convert scalar NA to Nullable instead of NullableArray

### CODE REVIEW

File: src/RCall.jl [hunk 1]
```diff
@@ -1,19 +1,19 @@
 __precompile__()
 module RCall
-using Compat, DataFrames, DataArrays
+using Compat, DataFrames, NullableArrays, CategoricalArrays
 import Compat.String
 
 import DataStructures: OrderedDict
 
-import Base: eltype, show, convert, isascii,
+import Base: eltype, show, convert, isascii, isnull,
```
Comment (nalimilan) [inline src/RCall.jl]: You shouldn't need `import` for this, merely `using` AFAICT.

File: test/dataframe.jl [hunk 2]
```diff
@@ -1,31 +1,52 @@
-using DataArrays,DataFrames
+using NullableArrays,CategoricalArrays,DataFrames
 
-v110 = rcopy(DataArray,reval("x <- 1:10"))
-@test isa(v110,DataVector)
-@test eltype(v110) == Cint
+@test rcopy(Nullable, RObject(1)).value == 1
+@test rcopy(Nullable, RObject("abc")).value == "abc"
+@test rcopy(RObject(Nullable(1))) == 1
+@test rcopy(Nullable, RObject(Nullable(1, true))).isnull
```
Comment (nalimilan) [inline test/dataframe.jl]: You should really avoid accessing the `isnull` field directly, as it's likely to go away soon (https://github.com/JuliaLang/julia/pull/18510). Use the accessor instead. (This also applies in other places.)

File: test/dataframe.jl [hunk 3]
```diff
@@ -1,31 +1,52 @@
...
+@test isa(station,NullableCategoricalArray)
 
-@test rcopy(DataArray,"c(NA,TRUE)").na == @data([NA,true]).na
-@test rcopy(DataArray,"c(NA,1)").na == @data([NA,1.0]).na
-@test rcopy(DataArray,"c(NA,1+0i)").na == @data([NA,1.0+0.0*im]).na
-@test rcopy(DataArray,"c(NA,1L)").na == @data([NA,one(Int32)]).na
-@test rcopy(DataArray,"c(NA,'NA')").na == @data([NA,"NA"]).na
-@test_throws ErrorException rcopy(DataArray,"as.factor(c('a','a','c'))")
-@test rcopy(PooledDataArray,"as.factor(c('a','a','c'))").pool == ["a","c"]
+@test rcopy(NullableArray,"c(NA,TRUE)").isnull == NullableArray([true,true], [true,false]).isnull
```
Comment (nalimilan) [inline test/dataframe.jl]: Same here for `NullableArray`. You can use `isequal` to test for equality of both contents and missingness, which is even better.

File: src/convert-data.jl [hunk 4]
```diff
@@ -39,12 +68,28 @@ function sexp(v::DataArray)
...
-    setattrib!(rv, Const.LevelsSymbol, sexp(v.pool))
-    setattrib!(rv, Const.ClassSymbol, sexp("factor"))
-    rv
+## CategoricalArray to sexp conversion.
+for typ in [:NullableCategoricalArray, :CategoricalArray]
+    @eval begin
+        function sexp{T<:Compat.String,N,R<:Integer}(v::$typ{T,N,R})
+            rv = protect(sexp(v.refs))
+            try
+                for (i,isna) = enumerate(v.refs .== 0)
```
Comment (nalimilan) [inline src/convert-data.jl]: `v.refs .== 0` will allocate a vector. Better do `== 0` inside the loop.

File: src/convert-data.jl [hunk 5]
```diff
@@ -39,12 +68,28 @@ function sexp(v::DataArray)
...
+                    if isna
+                        rv[i] = naeltype(eltype(rv))
+                    end
+                end
+                setattrib!(rv, Const.LevelsSymbol, sexp(v.pool.index))
```
Comment (nalimilan) [inline src/convert-data.jl]: Please use `CategoricalArrays.index(v)`, `ordered(v)` and `levels(v)`. Private fields should really not be accessed directly as they could change.

File: src/RCall.jl (hunk 1 above)
Comment (randy3k) [inline src/RCall.jl]: it is because RCall also has a isnull method.

File: src/convert-data.jl (hunk 5 above)
Comment (randy3k) [inline src/convert-data.jl]: `CategoricalArrays.index` does not work?

```julia
julia> v = CategoricalArray(repeat(["a", "b"], inner = 5));

julia> CategoricalArrays.levels(v)
2-element Array{String,1}:
 "a"
 "b"

julia> CategoricalArrays.index(v)
ERROR: MethodError: no method matching index(::CategoricalArrays.CategoricalArray{String,1,UInt32})
Closest candidates are:
  index(::CategoricalArrays.CategoricalPool{T,R<:Integer,V}) at /Users/Randy/.julia/v0.5/CategoricalArrays/src/pool.jl:161
```

File: src/convert-data.jl (hunk 4 above)
Comment (randy3k) [inline src/convert-data.jl]: thanks for catching it

File: test/dataframe.jl (hunk 3 above)
Comment (nalimilan) [inline test/dataframe.jl]: Yes, you need to do `NullableArray(1:3)`.

File: src/convert-data.jl (hunk 5 above)
Comment (nalimilan) [inline src/convert-data.jl]: Ah, indeed, at first I didn't intend it to be used outside of the package. For now the best approach is `CategoricalArrays.index(v.pool)`. I'll try to see if I can improve the API in next releases.

File: test/dataframe.jl (hunk 3 above)
Comment (randy3k) [inline test/dataframe.jl]: `isequal(NullableArray(1:3), 1:3)` or `isequal(NullableArray(1:3), collect(1:3))` both return false..

Comment (nalimilan) [inline test/dataframe.jl]: I meant isequal(rcopy(NullableArray,"c(NA,TRUE)"), NullableArray([true,true], [true,false])` and so on (i.e. convert both sides to a `NullableArray`).

**Pull request merged**